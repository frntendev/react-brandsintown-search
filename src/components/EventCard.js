import React from "react";
import PropTypes from "prop-types";
import { css } from "react-emotion";

const styles = {
  event: css`
    display: flex;
    min-height: 50px;
    background: #eeeeee85;
    margin-top: 10px;
    margin-bottom: 10px;
    align-items: center;
    padding: 5px;
    box-sizing: border-box;
    transition: 0.3s all ease;
    text-align: center;
    &:hover {
      box-shadow: 0 10px 20px rgba(0, 0, 0, 0.19), 0 6px 6px rgba(0, 0, 0, 0.23);
    }
    label: event-root;
  `,
  eventText: css`
    flex: 1;
    font-size: 1.2rem;
    color: #545454;
    label: eventText;
    @media (max-width: 758px) {
      font-size: 1rem;
    }
  `,
  eventLink: css`
    background: #48c5b6;
    color: #fff;
    text-decoration: none;
    font-size: 1.2rem;
    padding: 10px;
    white-space: pre;
    transition: 0.3s all ease;
    margin-left: 5px;
    cursor: pointer;
    label: eventLink;
    @media (max-width: 758px) {
      font-size: 1rem;
    }
    &:hover {
      background: #3fa99c;
    }
  `
};
const EventCard = props => {
  return (
    <div className={styles.event}>
      <span className={styles.eventText}>
        <strong>{props.country}</strong>
      </span>
      <span className={styles.eventText}>{props.city}</span>
      <span className={styles.eventText}>{props.date}</span>
      <span className={styles.eventLink} onClick={props.onSaveEvent}>
        Save
      </span>
      <a className={styles.eventLink} href={props.url} target="_blank">
        Book
      </a>
    </div>
  );
};

export default EventCard;

EventCard.defaultProps = {
  country: "Iran",
  city: "Tehran",
  date: "1-1-2018"
};

EventCard.propTypes = {
  /**
   * The country of an event
   */
  country: PropTypes.string.isRequired,
  /**
   * The city of an event
   */
  city: PropTypes.string.isRequired,
  /**
   * The date of an event
   */
  date: PropTypes.string.isRequired,
  /**
   * The url of an event
   */
  url: PropTypes.string,
  /**
   * A callback function that saves the favorite event
   */
  onSaveEvent: PropTypes.func
};
