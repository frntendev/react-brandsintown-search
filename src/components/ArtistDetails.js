import React from "react";
import PropTypes from "prop-types";
import { css } from "react-emotion";
import ArtistCard from "./ArtistCard";
import EventCard from "./EventCard";

const styles = {
  root: css`
    width: 80%;
    margin: 0 auto;
    label: root;
    @media (max-width: 758px) {
      width: 95%;
    }
  `
};

class ArtistDetails extends React.Component {
  handleSaveEvent = data => {
    this.props.onSaveEvent(data);
  };

  render() {
    const props = this.props;
    const { artist, events } = props.data;
    return (
      <div className={styles.root}>
        {artist && (
          <ArtistCard
            image={artist.image_url}
            name={artist.name}
            detailsInfo={[artist.facebook_page_url, artist.url, artist.id]}
            url={artist.url}
          />
        )}
        {events && (
          <div>
            {events.map((item, index) => {
              const id = item.id;
              const dateTime = new Date(item.datetime);
              const year = dateTime.getFullYear();
              const month = dateTime.getMonth() + 1;
              const day = dateTime.getDate();
              const date = `${year}-${month}-${day}`;
              const city = item.venue.city;
              const url = item.url;
              const country = item.venue.country;
              const currentEvent = {
                [id]: {
                  id: id,
                  name: artist.name,
                  imageUrl: artist.image_url,
                  country,
                  date,
                  city,
                  url
                }
              };
              return (
                <EventCard
                  key={`event-${index}`}
                  country={country}
                  city={city}
                  date={date}
                  url={url}
                  onSaveEvent={() => this.handleSaveEvent(currentEvent)}
                />
              );
            })}
          </div>
        )}
      </div>
    );
  }
}

export default ArtistDetails;

ArtistDetails.propTypes = {
  /**
   * A callback function to save the favorite event
   */
  onSaveEvent: PropTypes.func,
  /**
   * An object that contains artist details and events information
   * It contains `artist` and `events` objects
   */
  data: PropTypes.object.isRequired
};
