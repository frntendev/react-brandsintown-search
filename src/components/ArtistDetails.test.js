import React from "react";
import { shallow, mount, configure } from "enzyme";
import ArtistDetails from "./ArtistDetails";
import Adapter from "enzyme-adapter-react-16";
import EventCard from "./EventCard";
configure({ adapter: new Adapter() });

const initEvents = [
  {
    id: "21262663",
    artist_id: "90",
    url:
      "https://www.bandsintown.com/e/21262663?app_id=sd40fhp3731&came_from=267",
    on_sale_datetime: "2018-01-23T09:00:00",
    datetime: "2018-06-03T19:00:00",
    description: "",
    venue: {
      name: "Barclaycard Arena",
      latitude: "53.589234",
      longitude: "9.899232",
      city: "Hamburg",
      region: "04",
      country: "Germany"
    },
    offers: [
      {
        type: "Tickets",
        url:
          "https://www.bandsintown.com/t/21262663?app_id=sd40fhp3731&came_from=267",
        status: "available"
      }
    ],
    lineup: ["Shakira"]
  },
  {
    id: "19009960",
    artist_id: "90",
    url:
      "https://www.bandsintown.com/e/19009960?app_id=sd40fhp3731&came_from=267",
    on_sale_datetime: "2017-06-28T08:00:00",
    datetime: "2018-06-05T19:00:00",
    description: "Rescheduled from 8 Nov 2017",
    venue: {
      name: "LANXESS Arena",
      latitude: "50.9333333",
      longitude: "6.95",
      city: "Koln",
      region: "07",
      country: "Germany"
    },
    offers: [
      {
        type: "Tickets",
        url:
          "https://www.bandsintown.com/t/19009960?app_id=sd40fhp3731&came_from=267",
        status: "available"
      },
      {
        type: "VIP",
        url:
          "https://www.bandsintown.com/t/19009960?app_id=sd40fhp3731&came_from=267&custom=true",
        status: "available"
      }
    ],
    lineup: ["Shakira"]
  },
  {
    id: "19002149",
    artist_id: "90",
    url:
      "https://www.bandsintown.com/e/19002149?app_id=sd40fhp3731&came_from=267",
    on_sale_datetime: "",
    datetime: "2018-06-07T19:00:00",
    description: "Rescheduled from 12 Nov 2017",
    venue: {
      name: "Sportpaleis",
      latitude: "51.2301",
      longitude: "4.44156",
      city: "Merksem (Antwerpen)",
      region: "",
      country: "Belgium"
    },
    offers: [
      {
        type: "Tickets",
        url:
          "https://www.bandsintown.com/t/19002149?app_id=sd40fhp3731&came_from=267",
        status: "available"
      },
      {
        type: "VIP",
        url:
          "https://www.bandsintown.com/t/19002149?app_id=sd40fhp3731&came_from=267&custom=true",
        status: "available"
      }
    ],
    lineup: ["Shakira"]
  }
];
const initArtist = {
  id: "90",
  name: "Shakira",
  url: "https://www.bandsintown.com/a/90?came_from=267&app_id=sd40fhp3731",
  image_url: "https://s3.amazonaws.com/bit-photos/large/8518614.jpeg",
  thumb_url: "https://s3.amazonaws.com/bit-photos/thumb/8518614.jpeg",
  facebook_page_url: "https://www.facebook.com/shakira",
  mbid: "726cb350-d775-4e07-b006-898555f24b0d",
  tracker_count: 2728738,
  upcoming_event_count: 53
};
describe("ArtistDetails", () => {
  test("renders component itself correctly", () => {
    const renderedComponent = shallow(
      <ArtistDetails data={{ artist: {}, event: {} }} />
    );
    expect(renderedComponent).toMatchSnapshot();
  });

  test("renders component ArtistCard", () => {
    const renderedComponent = mount(<ArtistDetails data={{ artist: {} }} />);
    expect(renderedComponent.find("#artist-card").exists()).toEqual(true);
  });

  test("doesn't render component ArtistCard", () => {
    const renderedComponent = mount(<ArtistDetails data={{ event: {} }} />);
    expect(renderedComponent.find("#artist-card").exists()).toEqual(false);
  });

  test("render component Events", () => {
    const renderedComponent = shallow(
      <ArtistDetails
        data={{
          artist: initArtist,
          events: initEvents
        }}
      />
    );
    expect(renderedComponent.find(EventCard)).toHaveLength(3);
  });
});
