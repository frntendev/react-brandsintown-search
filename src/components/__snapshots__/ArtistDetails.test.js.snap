// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ArtistDetails renders component itself correctly 1`] = `
ShallowWrapper {
  "length": 1,
  Symbol(enzyme.__root__): [Circular],
  Symbol(enzyme.__unrendered__): <ArtistDetails
    data={
      Object {
        "artist": Object {},
        "event": Object {},
      }
    }
  />,
  Symbol(enzyme.__renderer__): Object {
    "batchedUpdates": [Function],
    "getNode": [Function],
    "render": [Function],
    "simulateEvent": [Function],
    "unmount": [Function],
  },
  Symbol(enzyme.__node__): Object {
    "instance": null,
    "key": undefined,
    "nodeType": "host",
    "props": Object {
      "children": Array [
        <ArtistCard
          detailsInfo={
            Array [
              undefined,
              undefined,
              undefined,
            ]
          }
          id="1"
          showClose={false}
        />,
        undefined,
      ],
      "className": "css-19fhf3m-root",
    },
    "ref": null,
    "rendered": Array [
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "function",
        "props": Object {
          "detailsInfo": Array [
            undefined,
            undefined,
            undefined,
          ],
          "id": "1",
          "image": undefined,
          "name": undefined,
          "showClose": false,
          "url": undefined,
        },
        "ref": null,
        "rendered": null,
        "type": [Function],
      },
      undefined,
    ],
    "type": "div",
  },
  Symbol(enzyme.__nodes__): Array [
    Object {
      "instance": null,
      "key": undefined,
      "nodeType": "host",
      "props": Object {
        "children": Array [
          <ArtistCard
            detailsInfo={
              Array [
                undefined,
                undefined,
                undefined,
              ]
            }
            id="1"
            showClose={false}
          />,
          undefined,
        ],
        "className": "css-19fhf3m-root",
      },
      "ref": null,
      "rendered": Array [
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "function",
          "props": Object {
            "detailsInfo": Array [
              undefined,
              undefined,
              undefined,
            ],
            "id": "1",
            "image": undefined,
            "name": undefined,
            "showClose": false,
            "url": undefined,
          },
          "ref": null,
          "rendered": null,
          "type": [Function],
        },
        undefined,
      ],
      "type": "div",
    },
  ],
  Symbol(enzyme.__options__): Object {
    "adapter": ReactSixteenAdapter {
      "options": Object {
        "enableComponentDidUpdateOnSetState": true,
      },
    },
  },
}
`;
